### PR TITLE
Add a `OperationWithResources` base class and custom operation support

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
In this PR we introduce a new class `OperationWithResources(qml.operations.Operation)`. This class instantiates an abstract method `resources(self) -> Resources:` which returns a `Resources` instance representing the custom resources for the operation. 

This will allow us to track resources for custom operations using `qml.specs()`

**Description of the Change:**
- Introduce `OperationWithResources`
- Modify CircuitGraph to account for custom depth
- Tests

### CODE REVIEW

File: pennylane/resource/resource.py [hunk 1]
```diff
@@ -44,7 +47,8 @@ class Resources:
         gates: 2
         depth: 2
         shots: 0
-        gate_types: {'Hadamard': 1, 'CNOT': 1}
+        gate_types:
+        {'Hadamard': 1, 'CNOT': 1}
```
Comment (Jaybsoni) [inline pennylane/resource/resource.py:51]: fixed a typo from before, previously did not match the terminal output.

File: pennylane/circuit_graph.py [hunk 2]
```diff
@@ -469,9 +471,42 @@ def get_depth(self):
...
+                self._extend_graph(self._operation_graph)
                 self._depth = rx.dag_longest_path_length(self._operation_graph) + 1
         return self._depth
 
+    def _extend_graph(self, graph: rx.PyDiGraph) -> rx.PyDiGraph:
+        """Extend graph to account for custom depth operations"""
+        custom_depth_node_dict = dict()
+        for op in self.operations:
+            if isinstance(op, OperationWithResources) and (d := op.resources().depth) > 1:
+                custom_depth_node_dict[graph.nodes().index(op)] = d
```
Comment (trbromley) [inline pennylane/circuit_graph.py:483]: I had questions about performance concerns for users outside of resource estimation workflows, but in that case `custom_depth_node_dict` would be empty, right?

File: pennylane/circuit_graph.py [hunk 3]
```diff
@@ -469,9 +471,42 @@ def get_depth(self):
                 self._operation_graph = self._graph.subgraph(
                     list(self._graph.nodes().index(node) for node in self.operations)
                 )
+                self._extend_graph(self._operation_graph)
                 self._depth = rx.dag_longest_path_length(self._operation_graph) + 1
         return self._depth
 
+    def _extend_graph(self, graph: rx.PyDiGraph) -> rx.PyDiGraph:
```
Comment (trbromley) [inline pennylane/circuit_graph.py:478]: I have a bit of caution adding a (private) method to `circuit_graph`. Could we instead have this functionality live in `resource`? I.e., if our circuit contains a `OperationWithResources`, then we call a `_extend_graph` private function which updates the underlying `PyDiGraph` before calculating its depth?

File: pennylane/resource/resource.py [hunk 4]
```diff
@@ -71,6 +75,35 @@ def _ipython_display_(self):
         print(str(self))
 
 
+class OperationWithResources(Operation):
```
Comment (trbromley) [inline pennylane/resource/resource.py:78]: To discuss the naming - I feel like `ResourceOp` is shorter and still descriptive.

**Review by trbromley — COMMENTED** [review]
Thanks @Jaybsoni! Great job on getting the resource estimation functionality together so quickly this week. I've left a few more comments.

I think we should not rush through this PR just to meet feature freeze week, let's give ourselves a bit more time next week before merging this in (i.e., so I think this is fine to be a feature for 0.31).

File: pennylane/circuit_graph.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/circuit_graph.py:483]: Correct! As long as there aren't any `OperationsWithResources` on the tape, the dict will remain empty and the circuit graph will be the same 👍🏼

File: pennylane/circuit_graph.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/circuit_graph.py:478]: I can move it to be a private method in resources. I don't know if that fits with best practices. How about as an alternative, it's not a private method in the class, but rather a private function in the file?

File: pennylane/resource/resource.py (hunk 4 above)
Comment (Jaybsoni) [reply pennylane/resource/resource.py:78]: I agree we should shorten it. To me `ResourceOp` sounds more like a new custom operation than a base class to inherit from.

Comment (DSGuala) [reply pennylane/resource/resource.py:78]: `ResourceOperation`?

Comment (soranjh) [reply pennylane/resource/resource.py:78]: `ResourceOperation` sounds good to me.

File: pennylane/circuit_graph.py [hunk 5]
```diff
@@ -469,9 +471,42 @@ def get_depth(self):
...
+                        "",
+                    )
+
+            # extend_graph:
+            def _link_graph(source_index, target_index):
+                if target_index == node_index:
+                    return sub_graph.nodes().index(f"{node_index}.{0}")
+                elif source_index == node_index:
+                    return sub_graph.nodes().index(f"{node_index}.{depth-1}")
+                return None
```
Comment (soranjh) [inline pennylane/circuit_graph.py:504]: Is this necessary?

File: tests/resource/test_resource.py [hunk 6]
```diff
@@ -131,13 +131,50 @@ def test_ipython_display(self, r, rep, capsys):
...
+        class CustomOpNoResource(OperationWithResources):
+            num_wires = 2
+
+        class CustomOPWithResources(OperationWithResources):
+            num_wires = 2
+
+            def resources(self):
+                return Resources(num_wires=self.num_wires)
+
+        with pytest.raises(TypeError, match="Can't instantiate abstract"):
```
Comment (soranjh) [inline tests/resource/test_resource.py:150]: The error message doesn't seem to match the original one.

File: pennylane/resource/resource.py [hunk 7]
```diff
@@ -71,6 +75,35 @@ def _ipython_display_(self):
...
+        shots: 0
+        gate_types:
+        {}
+        """
+        raise NotImplementedError(
+            f"Must define a resources method which returns a {Resources} instance."
+        )
+
+
 def _count_resources(tape, shots: int) -> Resources:
```
Comment (DSGuala) [inline pennylane/resource/resource.py:107]: When is this function called?

File: tests/resource/test_resource.py [hunk 8]
```diff
@@ -153,13 +190,39 @@ def _construct_tape_from_ops(lst_ops):
...
+        100,
+    ),
 )
 
 resources_data = (
     Resources(),
     Resources(2, 2, {"Hadamard": 1, "CNOT": 1}, 2, 0),
     Resources(3, 3, {"PauliZ": 1, "CNOT": 1, "RX": 1}, 2, 10),
     Resources(2, 6, {"Hadamard": 3, "RX": 2, "CNOT": 1}, 4, 100),
+    Resources(2, 5, {"Hadamard": 1, "CNOT": 1, "Identity": 1, "PauliZ": 2}, 5, 0),
```
Comment (soranjh) [inline tests/resource/test_resource.py:223]: Good to mention what these items are. i.e.,
`#  wires, gates, gate_types, depth, shots`

File: pennylane/resource/resource.py (hunk 4 above)
Comment (DSGuala) [inline pennylane/resource/resource.py:78]: Will `OperationWithResources` be top-level accessible? `qml.OperationWithResources`?

File: pennylane/resource/resource.py [hunk 9]
```diff
@@ -71,6 +75,35 @@ def _ipython_display_(self):
...
+            (:class:`~.Resources`): The resources required by this operation.
+
+        **Examples**
+
+        >>> class CustomOp(OperationWithResources):
+        ...     num_wires = 2
+        ...     def resources(self):
+        ...         return Resources(num_wires=self.num_wires, num_gates=3, depth=2)
+        ...
+        >>> op = CustomOp()
```
Comment (DSGuala) [inline pennylane/resource/resource.py:93]: Could be good to have an example like this in the docstring for `OperationWithResources`

**Review by DSGuala — COMMENTED** [review]
Left a few basic comments and questions.

File: pennylane/circuit_graph.py (hunk 5 above)
Comment (Jaybsoni) [reply pennylane/circuit_graph.py:504]: The last `return None`? It can be removed

File: pennylane/resource/resource.py (hunk 7 above)
Comment (Jaybsoni) [reply pennylane/resource/resource.py:107]: Internally called in `qml.specs`

File: pennylane/resource/resource.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/resource/resource.py:78]: I can add one here 👍🏼

Comment (Jaybsoni) [reply pennylane/resource/resource.py:78]: It currently isn't, I can make it top level accessible

File: pennylane/resource/resource.py (hunk 9 above)
Comment (Jaybsoni) [reply pennylane/resource/resource.py:93]: 👍🏼

**Comment by Jaybsoni** [comment]
This PR was reworked into a few smaller PRs, it is now obsolete

**Pull request closed without merging**